beluga_ros/include/beluga_ros/ndt_ellipsoid.hpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ROS_NDT_ELLIPSOID_HPP
+#define BELUGA_ROS_NDT_ELLIPSOID_HPP
+
+#include <beluga_ros/messages.hpp>
+
+
+/**
+ * \file
+ * \brief Utilities for NDT I/O over ROS interfaces.
+ */
+
+namespace beluga_ros {
+
+/// 
+template <typename MapType, int NDim>
+//beluga_ros::msg::MarkerArray& assign_obstacle_map(beluga::SparseValueGrid<MapType, NDim> grid){
+beluga_ros::msg::MarkerArray assign_obstacle_map(const beluga::SparseValueGrid<MapType, NDim>& grid){
+
+  // Get data from the grid
+  auto& map = grid.data();
+
+  beluga_ros::msg::MarkerArray message{};
+
+  int idCount = 0;
+  for (const auto& [key, cell] : map)
+  {
+    beluga_ros::msg::Marker marker;
+    marker.header.frame_id = "map";
+    marker.id = idCount++;  // TODO: Compile and check if this is the way to set id
+    marker.ns = "obstacles";
+    marker.type = beluga_ros::msg::Marker::SPHERE;
+    marker.action = beluga_ros::msg::Marker::ADD;
+
+    // Set the position
+    // TODO: check the mean's data type and how to access to its members
+    marker.pose.position.x = cell.mean[0];
+    marker.pose.position.y = cell.mean[1];
+    marker.pose.position.z = cell.mean[2];
+
+    // Set the scale based on the covariance
+    // TODO: check what values of the covariance should be used and if they need be changed before
```

Comment:
you need to set rotation and scale based on the covariance matrix.
For that you need to:

* Get eigenvectors of the matrix. You can use eigen for that, I cannot access the docs page for some reason, but I will copy links when it's back online.
* Make a matrix with each eigenvector as a column (maybe eigen provides that already).
  Order the columns so it's a rotation matrix (det == 1 and not det == -1).
* That's your rotation matrix, you can convert that to a quaternion for the ros message.
* The eigenvalues should be the x, y, z scale.

@hidmic is that reasonable to you?

Reply:
one thing, the eigenvectors solver may fail to converge to a solution, in that case we should have an alternative representation.
Probably, we should make it a cube of cell size, so the difference is obvious vs the ellipsoids.

But I don't think that should happen much in practice (or likely not at all) ....